kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/doubles/DoubleMatchersTest.kt

```diff
@@ -1,72 +1,1065 @@
+    }
+    
+    
+  }
+  
+  
+  private fun Double.toleranceValue(): Double {
+    return 5 * ulp
+  }
+  
+  private fun Double.slightlyGreater(): Double {
+    return this + (10 *ulp)
+  }
+  
+  private fun Double.muchGreater(): Double {
+    return this + (100 * ulp)
+  }
+  
+  private fun Double.slightlySmaller(): Double {
+    return this - (10 * ulp)
+  }
+  
+  private fun Double.muchSmaller(): Double {
+    return this - (100 * ulp)
+  }
+  
+  private fun shouldThrowAssertionError(message: String, vararg expression: () -> Any?) {
+    expression.forEach {
+      val exception = shouldThrow<AssertionError>(it)
+      exception.message shouldBe message
+    }
+  }
+  
+  private infix fun Double.shouldExactlyMatch(other: Double) {
+    this shouldBeExactly other
+    this shouldBe exactly(other)
+    this shouldThrowExceptionOnNotExactly other
+  }
+  
+  private infix fun Double.shouldThrowExceptionOnNotExactly(other: Double) {
+    shouldThrowAssertionError("$this should not equal $other",
+                              { this shouldNotBeExactly other },
+                              { this shouldNotBe exactly(other) }
+    )
+  }
+  
+  private infix fun Double.shouldNotMatchExactly(other: Double) {
+    this shouldNotBe exactly(other)
+    this shouldNotBeExactly other
+    this shouldThrowExceptionOnExactly other
+  }
+  
+  private infix fun Double.shouldThrowExceptionOnExactly(other: Double) {
+    shouldThrowAssertionError("$this is not equal to expected value $other",
+                              { this shouldBeExactly other },
+                              { this shouldBe exactly(other) }
+    )
+  }
+  
+  private fun Double.shouldMatchBetween(a: Double, b: Double, tolerance: Double) {
+    this.shouldBeBetween(a, b, tolerance)
+    this shouldBe between(a, b, tolerance)
+    
+    this.shouldThrowExceptionOnNotBetween(a, b, tolerance)
+  }
+  
+  private fun Double.shouldNotMatchBetween(a: Double, b: Double, tolerance: Double) {
+    this.shouldNotBeBetween(a, b, tolerance)
+    this shouldNotBe between(a, b, tolerance)
+    
+    this.shouldThrowExceptionOnBetween(a, b, tolerance)
+  }
+  
+  private fun Double.shouldThrowExceptionOnBetween(a: Double, b: Double, tolerance: Double) {
+    when {
+      this < a -> this.shouldThrowMinimumExceptionOnBetween(a, b, tolerance)
+      this > b -> this.shouldThrowMaximumExceptionOnBetween(a, b, tolerance)
+      else     -> throw IllegalStateException()
     }
+  }
+  
+  private fun Double.shouldThrowMinimumExceptionOnBetween(a: Double, b: Double, tolerance: Double) {
+    val message = "$this should be bigger than $a"
+    shouldThrowExceptionOnBetween(a, b, tolerance, message)
+  }
+  
+  private fun Double.shouldThrowMaximumExceptionOnBetween(a: Double, b: Double, tolerance: Double) {
+    val message = "$this should be smaller than $b"
+    shouldThrowExceptionOnBetween(a, b, tolerance, message)
+  }
+  
+  
+  private fun Double.shouldThrowExceptionOnBetween(a: Double,
```

Comment:
This is an extremely minor nit, but the kotlin style guide recommends formatting multi-line parameters like this:

```kotlin
private fun Double.shouldThrowExceptionOnBetween(
    a: Double,
    b: Double,
    tolerance: Double,
    message: String = "$this should be smaller than $b and bigger than $a"
) {
```

Which is nice in this case to avoid the long line due to the default parameter.